### PR TITLE
feat: add native LightGBM and CatBoost backend resolvers

### LINKED ISSUE #119: [FEATURE]: Add native backend resolvers for LightGBM and CatBoost
### Problem Statement
Currently, `trustlens/backends/` contains dedicated prediction resolvers for `sklearn.py` and `xgboost.py`. However, LightGBM and CatBoost are massive staples in tabular data science (and Kaggle competitions). Users currently have to manually extract probabilities and predictions to pass them into TrustLens, which defeats the library's "zero-friction" philosophy.

### Proposed Solution
Add `lightgbm.py` and `catboost.py` to the `trustlens/backends/` directory. 
- The LightGBM resolver must handle `lgb.Booster` vs `LGBMClassifier`.
- The CatBoost resolver must handle `catboost.Pool` and `CatBoostClassifier`, carefully managing categorical feature constraints when extracting probabilities.
- Both must map back to the standard `PredictionBundle`.

### Alternatives Considered
Forcing users to use `framework="manual"` and provide `y_prob` themselves. While functional, native support makes the tool significantly more attractive to the broader tabular ML community.

### Additional Context
This is a great issue for someone who wants to contribute to the core architectural parsing engine of TrustLens and is familiar with Gradient Boosting frameworks.

### PR DESCRIPTION
**Pull request by vaishnavidesai09**
## Summary

Adds native backend resolvers for LightGBM and CatBoost models, enabling automatic backend detection and prediction/probability extraction through the TrustLens backend registry.

## Related Issue

Closes #119

## Type of Change

- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 📝 Documentation update (changes to docs or examples)
- [ ] 🧹 Maintenance (refactoring, code cleanup, dependencies)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Changes Made

- [x] Added native LightGBM backend resolver with support for `LGBMClassifier` and `Booster`.
- [x] Added native CatBoost backend resolver with support for `CatBoostClassifier`.
- [x] Added automatic backend detection and resolver registration.
- [x] Added regression-model rejection logic for unsupported objectives.
- [x] Added integration and backend detection tests for both frameworks.

## Testing

- [x] All unit tests pass locally.
- [x] Added new tests for the changes made.

Commands run:

```bash
ruff check .
pytest tests/test_lightgbm.py -v
pytest tests/test_catboost.py -v
pytest -v
pre-commit run --all-files

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added support for CatBoost classifier models for end-to-end analysis.
  * Added support for LightGBM classifier models for end-to-end analysis.
  * Framework detection and prediction resolution now include CatBoost and LightGBM alongside existing sklearn and XGBoost support.
* **Documentation**
  * Documented native LightGBM & CatBoost backend availability under Unreleased changes.
* **Chores / Tests**
  * Improved test stability in headless environments.
  * Expanded backend coverage with new end-to-end and negative tests for regressors.
* **Refactor**
  * Streamlined Grad-CAM’s Matplotlib colormap imports for consistency.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

### CODE REVIEW

**Comment by vaishnavidesai09** [comment]
Hi! I've completed the implementation for #119.

Changes include:
- Native LightGBM backend resolver
- Native CatBoost backend resolver
- Automatic backend detection/registration
- Regression model rejection handling
- Comprehensive tests for detection, integration, manual overrides, and unsupported regressors

Validation performed:
- pre-commit run --all-files
- pytest tests/test_lightgbm.py -v
- pytest tests/test_catboost.py -v
- pytest -v

All tests pass locally (327 passed).

I'd appreciate a review when you have time. Thanks!

**Comment by vaishnavidesai09** [comment]
the CodeRabbit review feedback and pushed the updates to this PR.

Changes made:

Moved the LightGBM-specific import inside resolve() to preserve optional dependency handling.
Updated both LightGBM and CatBoost resolvers to derive y_pred from resolved probabilities when y_pred is not provided, ensuring consistency between y_pred and y_prob.

Validation rerun successfully:

pre-commit run --all-files
pytest tests/test_lightgbm.py -v
pytest tests/test_catboost.py -v
pytest -v

All tests pass locally (327 passed).

**Review by Khanz9664 — COMMENTED** [review]
@vaishnavidesai09 Thanks for the contribution and for taking on this feature.

I've reviewed the implementation and overall this is a strong addition to TrustLens. The resolver architecture is consistent with the existing backend pattern, framework detection and registration look correct, regression-model rejection handling is in place, and the test coverage is thorough.

I also reviewed the recent follow-up changes:

* The LightGBM optional-dependency import issue has been addressed.
* Prediction resolution now derives `y_pred` from the resolved probability matrix when appropriate, keeping `y_pred` and `y_prob` internally consistent across the new resolvers.

Before I approve, could you please add a changelog entry under the **Unreleased → Added** section describing the new native backend support for:

* LightGBM (`LGBMClassifier` and `Booster`)
* CatBoost (`CatBoostClassifier`)

Something along the lines of:

> **Native LightGBM & CatBoost Backends**: Added automatic backend detection and prediction resolution for LightGBM (`LGBMClassifier`, `Booster`) and CatBoost (`CatBoostClassifier`) models, including probability extraction, classification validation, and integration with the standard `PredictionBundle` pipeline.

Once that's added, I'll take another look and we should be in good shape.

Thanks again for the contribution 🚀

**Comment by Khanz9664** [comment]
@vaishnavidesai09 Before I can approve, we still need the failing CI checks addressed:

> CI / Linting

```
# Run linting and formatting:
pre-commit run --all-files

ruff check --fix

ruff format
```

Once the linting failures are resolved and CI is green, I'll do a final pass and we should be ready to merge.

**Comment by vaishnavidesai09** [comment]
Added the requested CHANGELOG entry under Unreleased → Added and resolved the follow-up review feedback. I've also rerun the validation suite locally:

* pre-commit run --all-files
* mypy .
* pytest

All tests are passing locally (327 passed). Thanks for the review!

**Review by Khanz9664 — COMMENTED** [review]
@vaishnavidesai09 Thank you for the contribution and for working through the review feedback so thoroughly.

I've completed my final review of the implementation. The core functionality requested in #119 is in place, the backend integrations are working as intended, the follow-up review comments have been addressed, the changelog entry has been added, and the test coverage for the new resolvers is comprehensive.

I can also see that you've been actively working through the remaining CI feedback. At this point, the only outstanding item is the linting workflow, and rather than sending you through another review cycle for what appears to be a minor repository-level cleanup, I'll take care of the remaining fix on my side after merging.

Thank you again for the contribution and for your patience throughout the review process. Native LightGBM and CatBoost support are valuable additions to TrustLens and help move the project closer to its goal of broad framework compatibility.

I'll handle the remaining cleanup and merge this shortly. 🚀

**Comment by Khanz9664** [comment]
@vaishnavidesai09 Merged. Thanks again for the contribution and for working through the review feedback. I handled the remaining linting cleanup in a follow-up maintainer PR after merge. The new LightGBM and CatBoost integrations are a great addition to TrustLens. 🚀
